cmd/mutate-image/main.go

```diff
@@ -0,0 +1,289 @@
+)
+
+// ExitError is an error which has an exit code to be used in os.Exit() to
+// return both an exit code and an error message
+type ExitError struct {
+	Code    int
+	Message string
+	Cause   error
+}
+
+func (e ExitError) Error() string {
+	return fmt.Sprintf("%s (exit code %d)", e.Message, e.Code)
+}
+
+// headerTransport sets headers on outgoing requests
+type headerTransport struct {
+	httpHeaders map[string]string
+	inner       http.RoundTripper
+}
+
+// RoundTrip implements http.RoundTripper
+func (ht *headerTransport) RoundTrip(in *http.Request) (*http.Response, error) {
+	for k, v := range ht.httpHeaders {
+		if http.CanonicalHeaderKey(k) == "User-Agent" {
+			// Docker sets this, which is annoying, since we're not docker.
+			// We might want to revisit completely ignoring this.
+			continue
+		}
+		in.Header.Set(k, v)
+	}
+
+	return ht.inner.RoundTrip(in)
+}
+
+type settings struct {
+	annotation,
+	label *[]string
+	image,
+	resultFileImageDigest,
+	resultFileImageSize string
+}
+
+func getAnnotation() []string {
+	var annotation []string
+
+	if flagValues.annotation != nil {
+		return append(annotation, *flagValues.annotation...)
+	}
+
+	return annotation
+}
+
+func getLabel() []string {
+	var label []string
+
+	if flagValues.annotation != nil {
+		return append(label, *flagValues.label...)
+	}
+
+	return label
+}
+
+var flagValues settings
+
+func initializeFlag() {
+	// Main flags for the image mutate step to define the configuration, for example
+	// the flag `image` will always be used.
+	pflag.StringVar(&flagValues.image, "image", "", "The name of image in container registry")
+	flagValues.annotation = pflag.StringArray("annotation", nil, "New annotations to add")
+	flagValues.label = pflag.StringArray("label", nil, "New labels to add")
+	pflag.StringVar(&flagValues.resultFileImageDigest, "result-file-image-digest", "", "A file to write the image digest to")
+	pflag.StringVar(&flagValues.resultFileImageSize, "result-file-image-size", "", "A file to write the image size to")
+}
+
+func main() {
+	if err := Execute(context.Background()); err != nil {
+		exitcode := 1
+
+		switch err := err.(type) {
+		case *ExitError:
+			exitcode = err.Code
+		}
+
+		log.Print(err.Error())
+		os.Exit(exitcode)
+	}
+}
+
+// Execute performs flag parsing, input validation and the image mutation
+func Execute(ctx context.Context) error {
+	initializeFlag()
+	pflag.Parse()
+
+	return runMutateImage(ctx)
+}
+
+func runMutateImage(ctx context.Context) error {
+	annotation := getAnnotation()
+	label := getLabel()
+
+	if flagValues.image == "" {
+		return &ExitError{Code: 100, Message: "the 'image' argument must not be empty"}
+	}
+
+	options := getOptions(ctx)
+	ref := flagValues.image
+
+	if len(annotation) != 0 {
+		desc, err := crane.Head(ref, *options...)
+		if err != nil {
+			return fmt.Errorf("checking %s: %v", ref, err)
+		}
+
+		if desc.MediaType.IsIndex() {
+			return fmt.Errorf("mutating annotations on an index is not yet supported")
```

Comment:
This might be a reason to prefer just using crane directly. If it doesn't support this already I suspect it will before we add support for it.